### PR TITLE
[ENH] Tafsut foundation model forecaster

### LINKED ISSUE #10993: [ENH] Tafsut forecasting foundation model
Another time series foundation model (with probabilistic predictions) that we could support:
https://github.com/Tafsut-FM/tafsut

Good first issue, adding a new model.

Guide on adding new estimators:
https://www.sktime.net/docs/developer-guide/add-estimators/

### PR DESCRIPTION
**Pull request by aryamanDutta**
#### Reference Issues/PRs

Fixes #10993.

#### What does this implement/fix? Explain your changes.

This PR introduces `TafsutForecaster`, an sktime adapter for the Tafsut univariate forecasting foundation model.

The forecaster provides:

- zero-shot probabilistic forecasting with the Tafsut pretrained model;
- point forecasts and quantile forecasts;
- support for sparse and absolute forecasting horizons;
- interpolation of requested quantiles from Tafsut's native quantile levels;
- support for missing values in the forecasting context;
- no support for exogenous variables or in-sample predictions;
- a small, locally initialized Tafsut configuration for estimator tests, avoiding a pretrained model download during testing.

The Tafsut model is loaded lazily as an optional dependency, and the loaded model is excluded from serialized estimator state.

#### Does your contribution introduce a new dependency? If yes, which one?

No new core dependencies are introduced. Tafsut is declared as a soft estimator dependency using the `python_dependencies` tag.

#### What should a reviewer concentrate their feedback on?

- The `TafsutForecaster` interface, particularly the handling of forecasting horizons and quantile outputs.
- The optional dependency and model-loading approach.
- Whether the estimator tags and test configuration are appropriate for Tafsut.

#### Did you add any tests for the change?

Yes. Added `sktime/forecasting/tests/test_tafsut.py`.

The tests cover:

- estimator tags and test parameters;
- point forecasts and sparse horizons;
- quantile forecasts;
- quantile ordering and interpolation;
- absolute horizons and missing context;
- quantile requests outside Tafsut's native quantile range.

The focused test suite passes:

```text
5 passed in 18.79s

### CODE REVIEW

File: docs/source/api_reference/forecasting.rst [hunk 1]
```diff
@@ -787,6 +787,14 @@ Domain agnostic foundation models
 
     TimerS1Forecaster
 
+.. currentmodule:: sktime.forecasting.tafsut
+
+.. autosummary::
+    :toctree: auto_generated/
+    :template: class.rst
+
+    TafsutForecaster
```
Comment (fkiraly) [inline docs/source/api_reference/forecasting.rst]: please order by alphabet

File: sktime/forecasting/tafsut.py [hunk 2]
```diff
@@ -0,0 +1,300 @@
...
+    >>> forecaster = TafsutForecaster(
+    ...     model_path=None, config=config, device="cpu"
+    ... )
+    >>> y = pd.Series(range(8))
+    >>> _ = forecaster.fit(y, fh=[1, 2, 3, 4])
+    >>> y_pred = forecaster.predict()
+    """
+
+    _tags = {
+        "authors": ["Tafsut-FM"],
```
Comment (fkiraly) [inline sktime/forecasting/tafsut.py]: these should be GitHub names of the developers, you can also add yours for the interface.

File: sktime/forecasting/tests/test_tafsut.py
Comment (geetu040) [inline sktime/forecasting/tests/test_tafsut.py:1]: we don't really need these tests, since these scenarios are already covered by `check_estimator`
if you want, you can write tests for quantiles (optional) following #10566 and golden outputs (must) following #10672

File: sktime/forecasting/tafsut.py [hunk 3]
```diff
@@ -0,0 +1,300 @@
+# copyright: sktime developers, BSD-3-Clause License (see LICENSE file)
+"""Tafsut foundation model forecaster."""
+
+__all__ = ["TafsutForecaster"]
+__author__ = ["Tafsut-FM"]
```
Comment (geetu040) [inline sktime/forecasting/tafsut.py]: keep the author's list consistent, it'd different in tags and you should add yourself as an author too since you are writing this class

File: sktime/forecasting/tafsut.py [hunk 4]
```diff
@@ -0,0 +1,300 @@
...
+
+    def __init__(
+        self,
+        model_path: str | None = "Tafsut-FM/tafsut-univariate-base",
+        config: dict | None = None,
+        device=None,
+        revision: str | None = None,
+        token: str | bool | None = None,
+        cache_dir: str | None = None,
+        local_files_only: bool = False,
```
Comment (geetu040) [inline sktime/forecasting/tafsut.py]: we should avoid these 3 parameters entirely:
- since they are not followed across other hf based models
- we can still configure them using env vars or hf-level config
- they are not so useful to be explicitly mentioned here as `__init__` arguments and it's cleaner to do this as the way above

File: sktime/forecasting/tafsut.py [hunk 5]
```diff
@@ -0,0 +1,300 @@
...
+            horizon=horizon,
+        )
+
+    def _get_relative_horizon(self, fh):
+        relative = fh.to_relative(self.cutoff).to_numpy(dtype=int)
+        if len(relative) == 0 or np.any(relative <= 0):
+            raise NotImplementedError(
+                "TafsutForecaster does not support in-sample predictions."
+            )
+        return relative
```
Comment (geetu040) [inline sktime/forecasting/tafsut.py:152]: same as the comment for `_to_numpy` below

File: sktime/forecasting/tafsut.py [hunk 6]
```diff
@@ -0,0 +1,300 @@
...
+            raise ValueError("Tafsut model must define a non-empty quantile sequence.")
+        if np.any(np.diff(quantiles) <= 0):
+            raise ValueError("Tafsut model quantiles must be strictly increasing.")
+        return quantiles
+
+    @staticmethod
+    def _to_numpy(values):
+        if hasattr(values, "detach"):
+            values = values.detach().cpu().numpy()
+        return np.asarray(values)
```
Comment (geetu040) [inline sktime/forecasting/tafsut.py]: we are only calling this once and it can be a single when we know where are using it, so a util function for this doesn't really add value

File: sktime/forecasting/tafsut.py [hunk 7]
```diff
@@ -0,0 +1,300 @@
...
+
+    def _get_forecast_values(self, fh):
```
Comment (geetu040) [inline sktime/forecasting/tafsut.py:163]: it is hard to follow what is exactly happening with so many small utility functions
better to keep the common behaviour from `_predict` and `_predict_quantiles` in just one simple function instead of breaking that further down into very small hard to follow functions.

also the naming can be improved like `_get_quantiles` makes us think you are going to do a forward pass and collect the quantile outputs but that's not the case, it rather returns the hard-coded quantiles allowed values from config

File: sktime/forecasting/tafsut.py [hunk 8]
```diff
@@ -0,0 +1,300 @@
...
+            [
+                np.array(
+                    [
+                        np.interp(requested_quantile, native_quantiles, row)
+                        for row in values
+                    ]
+                )
+                for requested_quantile in requested
+            ]
+        )
```
Comment (geetu040) [inline sktime/forecasting/tafsut.py:206]: this would be another case of inconsistent `predict_proba`, could you please implement custom `_predict_proba` as discussed with other estimators in #10566

File: sktime/forecasting/tafsut.py [hunk 9]
```diff
@@ -0,0 +1,300 @@
...
+            else:
+                self.model = TafsutModel.from_pretrained(
+                    self.model_path,
+                    device=self.device,
+                    revision=self.revision,
+                    token=self.token,
+                    cache_dir=self.cache_dir,
+                    local_files_only=self.local_files_only,
+                )
+        return self.model
```
Comment (geetu040) [inline sktime/forecasting/tafsut.py]: this function can be refactored as following
```suggestion
    def load(self):
        if self.model is not None:
            return self.model

        from tafsut import TafsutConfig, TafsutModel

        if self.model_path is not None:
            return TafsutModel.from_pretrained(
                self.model_path,
                device=self.device,
                revision=self.revision,
                token=self.token,
                cache_dir=self.cache_dir,
                local_files_only=self.local_files_only,
            )
        
        import torch

        config = TafsutConfig(**(self.config or {}))
        device = self.device
        if device is None:
            device = "cuda" if torch.cuda.is_available() else "cpu"
        device = torch.device(device)
        if device.type == "cuda" and not torch.cuda.is_available():
            raise RuntimeError("CUDA was requested but is not available")
        self.model = TafsutModel(config).to(device)
        self.model.eval()
```

**Review by geetu040 — CHANGES_REQUESTED** [review]
Thanks for the PR and explaining the changes and decisions thoroughly. This is a very well communicated work. I agree with the implementation and the workflow and have left some comments for improvements below.

**Comment by aryamanDutta** [comment]
@geetu040 Thanks for the feedback!..i'll shortly go through the requested changes and make the necessary changes.

File: sktime/forecasting/tafsut.py (hunk 4 above)
Comment (geetu040) [reply sktime/forecasting/tafsut.py]: I don't have a strong opposition against this but I would prefer to remove `local_files_only` since it's not really useful for this particular model

**Comment by aryamanDutta** [comment]
@geetu040 @fkiraly  Thanks a lot for the detailed review and for the suggestions!

I've addressed all of the requested changes:

removed the redundant bespoke tests;
updated the author metadata;
removed the explicit Hugging Face loading parameters;
refactored the model loading/caching into load();
removed _to_numpy() and simplified the prediction helpers;
renamed _get_quantiles() to _get_allowed_quantiles();
implemented the custom _predict_proba() using HistogramQPD;
added the required second get_test_params() configuration and ensured both configurations work correctly.

I also ran the full check_estimator suite for TafsutForecaster, including the deepcopy/serialization and probabilistic prediction checks, and all checks pass.

Let me know if any further changes are needed.

**Review by fkiraly — APPROVED** [review]
Looks good now!
